### ISSUE TITLE
Should this be followed using jupyter notebook, spyder or interactively?

### ISSUE THREAD

**Issue by gcapes** [datacarpentry/python-ecology-lesson#364]
These options are all mentioned in the setup episode, but no instruction is given. The output suggests this should be followed interactively, given `print` isn't required to print a variable in many of the examples.

Perhaps explicit instruction should be given how to follow the notes, or the notes should be written such that they can be followed equally whichever choice is made?

**Comment by wrightaprilm** [comment]
This is sort of the perpetual thorn in our side. There are so many more ways to get at Python than many of the other lesson languages. The instructions we use are shared with SWC [data and plotting](http://swcarpentry.github.io/python-novice-gapminder/setup/) lesson and the [dc-py-es](https://datacarpentry.org/python-ecology-lesson-es/) lesson. The inflammation lesson uses [these](http://swcarpentry.github.io/python-novice-inflammation/setup/) ones (maintained by co-maintainer on this repo @maxim-belkin). A while back @katrintirok pushed some more comprehensive instructions for Spyder. There have been various efforts to integrate a more thorough look at the Jupyter Notebook.

We've always tried, to some extent, to be agnostic on the exact delivery mechanism. I would love to have more flavors of set-up instructions, or a bank of setup instructions, that instructors could pull from as they are assembling the workshop website. The ideal case, in my opinion, would be that these instructions could be shared across lesson programs (SWC, DC, LC). I wouldn't want to maintain them, but maintainers should be looped in on issues affecting the changes.

Basically, having more detailed and varied set-up instructions has been getting kicked between each lesson and each lesson organization for years. Since this issue is popping up in my backyward this go-around, I'm willing to facilitate discussion on it, and to try to connect folks who are all thinking about this, but I really don't want to and don't have the skillset to lead on this.

Since this is the aforementioned perpetual thorn, I'm tagging in some folks who I know have thought about installs in the past. @fmichonneau, @ethanwhite, @rgaiacs. If we are going to make some sort of sustained push for more thorough getting started instructions, we'd want to set some sort of timeline for that, possibly in conjunction with the summer lesson releases.

**Comment by ethanwhite** [comment]
I don't have time to contribute meaningful work on this at the moment, but happy to provide historical context on specific things as you go if that proves to be useful. As a starting point I'll say that the initial work we did on making installation and setup much more stable and straightforward in 2012 was I think critical to the long-term success of the Carpentries. I think any time that folks have to dedicate to this will be super valuable for learners and instructors alike.

**Comment by katrinleinweber** [comment]
Has anyone else been mulling over [Joel Grus' "I don't like notebooks" talk](https://twitter.com/joelgrus/status/1044318416012750850)? Maybe recommending Spyder is the best compromise, because it has all the interactivity that JNs have (plus inline plots) but it's also a bridge towards other IDEs like [PyCharm (Community or Edu even has interactive tutorials for further self-study)](https://www.jetbrains.com/education/download/#section=pycharm-edu).

**Comment by maxim-belkin** [comment]
Yes, I saw that talk and even invited Joel to talk about it more some time ago... (got no response). I have no objections to listing Spyder on the same page as Jupyter Notebooks.

**Comment by katrinleinweber** [comment]
> got no response

I presume he said what he wanted to say in his talk ;-)

[Spyder started pushing an extension](https://www.spyder-ide.org/blog/spyder-kite-funding/) along with its installer, though. Reeks of adware, but it seems to emulate what PyCharm has built in. I'm beginning to understand why this issue is so tricky *sigh*

**Comment by maxim-belkin** [comment]
> > got no response
> 
> I presume he said what he wanted to say in his talk ;-)

There is nothing wrong with spreading worthy ideas. In his talk, Joel highlights a few issues that are important for any IDE out there, Spyder included.

> [Spyder started pushing an extension](https://www.spyder-ide.org/blog/spyder-kite-funding/) along with its installer, though. Reeks of adware, but it seems to emulate what PyCharm has built in. I'm beginning to understand why this issue is so tricky _sigh_

I don't see that extension in Spyder as a problem that should prevent us from listing it as an alternative IDE.

**Comment by katrinleinweber** [comment]
Here's a work-around for the most annoying issue in PyCharm with IPython (plots not easily viewable): https://github.com/TIBHannover/python-ecology-lesson/commit/427e91d8aba90ea5c4003dc90d58e5112ab70e12

**Comment by maxim-belkin** [comment]
This is one of the things that has to change or has to be configurable in Jupyter Notebooks (not in PyCharm): in Jupyter notebooks plots are shown automatically. This is not the default behavior of matplotlib and one has to call `matplotlib.pyplot.show()`. It simplifies lives for programmers who know about it already and can appreciate that they don't need to type yet another function call. But it confuses those who are just getting started with Python.

`_` variable is the other thing that doesn't work in Jupyter Notebooks...

**Comment by mlangseth** [comment]
I agree that offering flexibility in what technology instructors choose to use is helpful. I am wondering if it would be a good idea to use "Jupyter Notebook" instead of "IPython Notebook"? I am specifically referring to Episode 02 - Starting with Data: "If you are working in **IPython Notebook** be sure that you start your notebook in the workshop directory."

**Comment by maxim-belkin** [comment]
> I am wondering if it would be a good idea to use "Jupyter Notebook" instead of "IPython Notebook"? I am specifically referring to Episode 02 - Starting with Data: "If you are working in **IPython Notebook** be sure that you start your notebook in the workshop directory."

Yes, this is a good idea, @mlangseth! I'd appreciate if you could submit a PR to fix that.
